### ISSUE TITLE
[Usage]: Is it compatible with the original π 0.5？

### ISSUE THREAD

**Issue by Gulan1** [RLinf/RLinf#498]
### Problem description

Hello, thank you very much for your continuous open source work. May I ask if the fine-tuning of π 0.5 in your project is compatible with the original π 0.5. So can I use the model trained with the original π 0.5 to directly fine tune it?

### Before submitting a new issue...

- [x] Have you checked relevant issues, FAQs (https://rlinf.readthedocs.io/en/latest/rst_source/faq.html), or asked the chatbot at the top right corner of the [documentation page](https://rlinf.readthedocs.io).

**Comment by chenkang455** [comment]
Hi, @Gulan1, we built our RL model based on the official pi model (as seen in [here](https://github.com/RLinf/RLinf/blob/d34f91b62f7817187122e83dd3bf4b31104c6f68/rlinf/models/embodiment/openpi_action_model.py#L68)). By retaining most of the original functions and integrating RL-related algorithms, our model remains **fully compatible with the official version**.

This means you can:

- Train a standard SFT model using the official code and load it into our framework.

- Load a model trained with our RL pipeline into the official pi model for evaluation.

Both workflows are seamlessly supported.

**Comment by Gulan1** [comment]
Hello, @chenkang455 thank you very much for your patient reply
Fine tuning the π 0.5 real-time environment interaction part can be changed to the state where I am using an offline dataset (including action, reward, done, image). Can the PPO algorithm in the current code be fine tuned using these offline data?

**Comment by chenkang455** [comment]
Hi @Gulan1, our current codebase is designed specifically for online on-policy algorithms like PPO/GRPO. If you intend to use offline data for RL tuning, you should look into offline RL algorithms. On-policy algorithms like PPO require continuous interaction with the environment to collect fresh data for optimal performance.

The offline RL pipeline is similar to the SFT process. Fortunately, we have supported the sft process in RLinf, and you can modify the codebase yourself to support offline RL, which is also a future direction we are exploring.

**Comment by Gulan1** [comment]
> 你好[@Gulan1](https://github.com/Gulan1)我们目前的代码库是专门为在线策略算法（例如 PPO/GRPO）设计的。如果您打算使用离线数据进行强化学习调优，则应该考虑离线强化学习算法。像 PPO 这样的策略算法需要与环境持续交互以收集最新数据，从而达到最佳性能。
> 
> 离线强化学习流程与SFT流程类似。幸运的是，RLinf已经支持SFT流程，您可以自行修改代码库以支持离线强化学习，这也是我们正在探索的未来方向。

Hello, @chenkang455 thank you very much for your patient reply
I have already used the SFT process you mentioned for offline learning, but after checking, the generated model architecture seems to be different from the original π trained model. So, how should I proceed with inference

**Comment by chenkang455** [comment]
@Gulan1 Hi, can you elaborate it more precisely? Do you mean you cannot load our trained checkpoint into the official pi0 pytorch code? From my experiment, it is compatible. Can you provide me more info? thx a lot.

**Comment by Gulan1** [comment]
> [@Gulan1](https://github.com/Gulan1)您好，能否详细说明一下？您的意思是无法将我们训练好的检查点加载到官方的 Pi0 PyTorch 代码中吗？根据我的实验，它是兼容的。能否提供更多信息？非常感谢。

<img width="1560" height="336" alt="Image" src="https://github.com/user-attachments/assets/3fb453e0-1a7f-44f8-bf8f-1aea22813513" />

<img width="1032" height="298" alt="Image" src="https://github.com/user-attachments/assets/2a0ad215-4d94-401b-afbf-1e53481a8b5d" />                                      

@chenkang455 Hello, thank you for your response
The two pictures above are the model folders obtained from 800 rounds of offline training using the SFT process. Is it normal for his checkpoint to be divided into three, and then I tried using π PyTorch inference method to add a loading distribution model that seems to be able to reason, but the action is very strange? Is this reasonable?

**Comment by chenkang455** [comment]
@Gulan1 Thanks for providing that! We provide the script for eval with original openpi models in https://github.com/chenkang455/RLinf/tree/feature/add_pi_eval/toolkits/eval_scripts_openpi. Can you try this script for another try?

Besides, I use the merge script provided below for merging 4 safetensors:

``` python
"""
Merge sharded safetensors files into a single file
"""
import json
from pathlib import Path
from safetensors import safe_open
from safetensors.torch import save_file
import torch
from tqdm import tqdm


def merge_safetensors(checkpoint_dir: str, output_file: str = None):
    """
    Merge sharded safetensors files
    
    Args:
        checkpoint_dir: checkpoint directory path
        output_file: output file path, if None then use checkpoint_dir/model.safetensors
    """
    checkpoint_dir = Path(checkpoint_dir)
    
    # Read index.json
    index_file = checkpoint_dir / "model.safetensors.index.json"
    if not index_file.exists():
        raise FileNotFoundError(f"Index file not found: {index_file}")
    
    with open(index_file, 'r') as f:
        index_data = json.load(f)
    
    weight_map = index_data["weight_map"]
    
    # Determine output file path
    if output_file is None:
        output_file = checkpoint_dir / "model.safetensors"
    else:
        output_file = Path(output_file)
    
    print(f"Starting to merge safetensors files...")
    print(f"Input directory: {checkpoint_dir}")
    print(f"Output file: {output_file}")
    
    # Collect all files that need to be loaded
    shard_files = set(weight_map.values())
    print(f"Found {len(shard_files)} shard files")
    
    # Load all weights
    merged_state_dict = {}
    
    for shard_file in tqdm(sorted(shard_files), desc="Loading shards"):
        shard_path = checkpoint_dir / shard_file
        if not shard_path.exists():
            raise FileNotFoundError(f"Shard file not found: {shard_path}")
        
        with safe_open(str(shard_path), framework="pt", device="cpu") as f:
            # Get all keys in this shard
            keys = f.keys()
            for key in keys:
                if key in merged_state_dict:
                    print(f"Warning: key {key} is duplicated, will overwrite")
                merged_state_dict[key] = f.get_tensor(key)
    
    print(f"Loaded {len(merged_state_dict)} weights in total")
    print(f"Saving merged model to {output_file}...")
    
    # Save merged model
    save_file(merged_state_dict, str(output_file))
    
    print(f"Done! Merged model saved to: {output_file}")
    print(f"File size: {output_file.stat().st_size / (1024**3):.2f} GB")


if __name__ == "__main__":
    import argparse
    
    parser = argparse.ArgumentParser(description="Merge sharded safetensors files")
    parser.add_argument(
        "--checkpoint_dir",
        type=str,
        required=True,
        help="checkpoint directory path (contains model-*-of-*.safetensors and index.json)"
    )
    parser.add_argument(
        "--output_file",
        type=str,
        default=None,
        help="output file path (default is checkpoint_dir/model.safetensors)"
    )
    
    args = parser.parse_args()
    merge_safetensors(args.checkpoint_dir, args.output_file)
```

**Comment by chenkang455** [comment]
If these can not solve the problem, can you pull the newest code and re-try it? We previously found a small bug in SFT process, but we are not sure your current code version, thx a lot!
